app/services/tag_hashtag_data_source.rb

```diff
@@ -55,4 +55,36 @@ def self.search(guardian, term, limit)
   def self.search_sort(search_results, _)
     search_results.sort_by { |result| result.text.downcase }
   end
+
+  def self.search_without_term(guardian, limit)
+    return [] if !SiteSetting.tagging_enabled
+
+    tags = DB.query(<<~SQL, limit: limit).map(&:name)
+    SELECT tags.name, MAX(posts.created_at)
+    FROM tags
+    INNER JOIN topic_tags ON topic_tags.tag_id = tags.id
+          INNER JOIN topics ON topic_tags.topic_id = topics.id
+          INNER JOIN posts ON topics.id = posts.topic_id
+          WHERE posts.deleted_at IS NULL AND topics.deleted_at IS NULL
+            AND posts.created_at > (NOW() - INTERVAL '2 WEEKS')
+    GROUP BY tags.name
+    ORDER BY MAX(posts.created_at) DESC
+    LIMIT :limit
+    SQL
+
+    tags_with_counts, _ =
+      DiscourseTagging.filter_allowed_tags(
+        guardian,
+        with_context: true,
+        limit: limit,
+        for_input: true,
+        order_search_results: true,
+        only_tag_names: tags,
+      )
+
+    TagsController
```

Comment:
This is kind of a code smell here when we're reaching out to a class. Perhaps we should be using a serializer here instead?

Reply:
Yeah I really don't like this is in a controller either, I am not really sure why it is. It's used here too:

https://github.com/discourse/discourse/blob/c1ffa4e7b39ad66600440f63c19a6aa7cc2a9f77/app/serializers/detailed_tag_serializer.rb#L8-L10

I will refactor it but I would prefer to do it in a standalone PR